sklearn/svm/_libsvm_sparse.pyx

```diff
@@ -155,59 +156,61 @@ def libsvm_sparse_train (int n_features,
     cdef cnp.npy_intp SV_len = get_l(model)
     cdef cnp.npy_intp n_class = get_nr(model)
 
-    cdef cnp.ndarray[int, ndim=1, mode='c'] n_iter
+    cdef int[::1] n_iter
     n_iter = np.empty(max(1, n_class * (n_class - 1) // 2), dtype=np.intc)
-    copy_n_iter(n_iter.data, model)
+    copy_n_iter(<char *> &n_iter[0], model)
 
     # copy model.sv_coef
     # we create a new array instead of resizing, otherwise
     # it would not erase previous information
-    cdef cnp.ndarray sv_coef_data
+    cdef cnp.float64_t[::1] sv_coef_data
     sv_coef_data = np.empty((n_class-1)*SV_len, dtype=np.float64)
-    copy_sv_coef (sv_coef_data.data, model)
+    copy_sv_coef (<char *> &sv_coef_data[0] if sv_coef_data.size > 0 else NULL, model)
 
-    cdef cnp.ndarray[cnp.int32_t, ndim=1, mode='c'] support
+    cdef cnp.int32_t[::1] support
     support = np.empty(SV_len, dtype=np.int32)
-    copy_support(support.data, model)
+    copy_support(<char *> &support[0] if support.size > 0 else NULL, model)
 
     # copy model.rho into the intercept
     # the intercept is just model.rho but with sign changed
-    cdef cnp.ndarray intercept
+    cdef cnp.float64_t[::1]intercept
     intercept = np.empty(n_class*(n_class-1)//2, dtype=np.float64)
-    copy_intercept (intercept.data, model, intercept.shape)
+    copy_intercept (<char *> &intercept[0], model, <cnp.npy_intp *> intercept.shape)
 
     # copy model.SV
     # we erase any previous information in SV
     # TODO: custom kernel
     cdef cnp.npy_intp nonzero_SV
     nonzero_SV = get_nonzero_SV (model)
 
-    cdef cnp.ndarray SV_data, SV_indices, SV_indptr
+    cdef cnp.float64_t[::1] SV_data
+    cdef cnp.int32_t[::1] SV_indices, SV_indptr
     SV_data = np.empty(nonzero_SV, dtype=np.float64)
     SV_indices = np.empty(nonzero_SV, dtype=np.int32)
     SV_indptr = np.empty(<cnp.npy_intp>SV_len + 1, dtype=np.int32)
-    csr_copy_SV(SV_data.data, SV_indices.shape, SV_indices.data,
-                SV_indptr.shape, SV_indptr.data, model, n_features)
+    csr_copy_SV(<char *> &SV_data[0] if SV_data.size > 0 else NULL, <cnp.npy_intp *> SV_indices.shape,
+                <char *> &SV_indices[0] if SV_indices.size > 0 else NULL, <cnp.npy_intp *> SV_indptr.shape,
+                <char *> &SV_indptr[0] if SV_indptr.size > 0 else NULL, model, n_features)
```

Comment:
I actually added the if conditions because the tests were failing because of index errors related to probA and probB. Therefore I also applied this condition to other places just to be on the safe side, so that no unexpected errors occur. I am okay to apply the condition to only those variables which fail the tests and remove it from the others if you feel that would be suitable.

Reply:
Also just to clarify the tests pass generally on my local mac as well as most environments but fail on the windows build and one of the ubuntu builds.

Reply:
> I actually added the if conditions because the tests were failing because of index errors related to probA and probB.

For `probA` and `probB` it makes sense because they can be defined as `np.empty(0)` which is 0-D array and thus the need for your trick.

> Therefore I also applied this condition to other places just to be on the safe side, so that no unexpected errors occur.

I am not sure this is wise because reading the code, I would have the same expectations as for the `probA` and `probB` and thus misinterpret the code. This would be misleading. I would be more in favour of limiting to only the case where we get the failure and check that we don't have any initialization that uses 0-D array.

Reply:
> Also just to clarify the tests pass generally on my local mac as well as most environments but fail on the windows build and one of the ubuntu builds.

Certainly, we end up in some undefined behaviours that could therefore change from one compiler to another.

Reply:
Okay I'll try it out with only the obvious required variables and see how the CI goes.